### PR TITLE
implement programmatic default providers based on context in the engine and Go

### PR DESCRIPTION
**Pull request by tgummerer**
This PR takes an alternative approach for implementing programmatic default providers, that was suggested in the internal design doc.

Instead of making the default provider global, we create an API with a callback that creates a new subcontext.  In that subcontext and only in that subcontext is the provider set as the default provider.  We achieve this by setting up a new monitor with a different port, so the engine can distinguish between the contexts.

/xref https://github.com/pulumi/pulumi/issues/2059
Fixes https://github.com/pulumi/pulumi/issues/16169
Fixes https://github.com/pulumi/pulumi/issues/16163

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: proto/pulumi/resource.proto [hunk 1]
```diff
@@ -238,3 +239,11 @@ message TransformResponse {
     google.protobuf.Struct properties = 1; // the transformed input properties.
     TransformResourceOptions options = 2; // the options for the resource.
 }
+
+message RegisterDefaultProvidersRequest {
```
Comment (Frassle) [inline proto/pulumi/resource.proto:243]: I wonder if we should name this something more general? 
We could use the same trick for something like contextual parents as well.

File: proto/pulumi/resource.proto [hunk 2]
```diff
@@ -238,3 +239,11 @@ message TransformResponse {
...
+    repeated string providers = 1; // the provider refs to register as the default providers.
+}
+
+message RegisterDefaultProvidersResponse {
+    string monitorPort = 1;
```
Comment (Frassle) [inline proto/pulumi/resource.proto:248]: I'd lean towards having this return the full grpc target, not just the port. Don't assume localhost in the SDKs.

Comment (Frassle) [reply proto/pulumi/resource.proto:248]: That matches what we do for things like loaders and mappers.

File: pkg/resource/deploy/source_eval.go [hunk 3]
```diff
@@ -419,6 +422,12 @@ func (d *defaultProviders) handleRequest(req providers.ProviderRequest) (provide
 		return ref, nil
 	}
 
+	if d.explicitDefaultProviders[req.Package().Name().String()] {
+		logging.V(9).Infof("package %s has an explicit default provider.  Denying creation of an implicit one",
```
Comment (Frassle) [inline pkg/resource/deploy/source_eval.go:426]: I wouldn't have thought we'd ever get here? If a default already exists shouldn't we just be picking up that default, not triggering a registerDefaultProviderEvent?

File: proto/pulumi/resource.proto (hunk 1 above)
Comment (tgummerer) [reply proto/pulumi/resource.proto:243]: Yeah, the same trick could work in other places. Should the parameters then be a union type instead?  Or would it make sense to just have different grpc methods for it?  Maybe `CreateNewContext` would work for a unionized call?

File: proto/pulumi/resource.proto (hunk 2 above)
Comment (tgummerer) [reply proto/pulumi/resource.proto:248]: Makes sense, will change that.

File: pkg/resource/deploy/source_eval.go (hunk 3 above)
Comment (tgummerer) [reply pkg/resource/deploy/source_eval.go:426]: Yeah, I think this made more sense in the previous iteration without the callback based approach, I don't think this is necessary anymore.

File: proto/pulumi/resource.proto (hunk 1 above)
Comment (Frassle) [reply proto/pulumi/resource.proto:243]: I think just a message where all the fields are optional (or zero) would work fine.

File: proto/pulumi/resource.proto [hunk 4]
```diff
@@ -250,5 +254,13 @@ message RegisterProviderRequest {
...
\ No newline at end of file
+    string ref = 1;
+}
+
+message CreateNewContextRequest {
+    repeated string providers = 1; // the provider refs to register as the default providers.
+}
+
+message CreateNewContextResponse {
+    string monitorAddr = 1;
```
Comment (Frassle) [inline proto/pulumi/resource.proto:265]:
```suggestion
    string monitor_target = 1;
```
proto files ought to be snake_case for field names (we we're inconsistent about this before). Also we're trying to call these grpc address "target", that matches the ones for the loader and mapper.

File: pkg/resource/deploy/source_eval.go [hunk 5]
```diff
@@ -453,6 +454,10 @@ func (d *defaultProviders) handleRequest(req providers.ProviderRequest) (provide
 	return ref, nil
 }
 
+func (d *defaultProviders) RegisterDefaultProvider(provider providers.Reference) {
+	d.providers[provider.String()] = provider
```
Comment (Frassle) [inline pkg/resource/deploy/source_eval.go:458]: Is provider.String() right here? if yes why don't we reuse that logic in setProviderAsDefault?

File: pkg/resource/deploy/source_eval.go [hunk 6]
```diff
@@ -2165,6 +2259,73 @@ type registerResourceEvent struct {
...
+	d := &defaultProviders{
+		defaultProviderInfo: rm.defaultProviders.defaultProviderInfo,
+		providers:           deepcopy.Copy(rm.defaultProviders.providers).(map[string]providers.Reference),
+		config:              rm.defaultProviders.config,
+		requests:            make(chan defaultProviderRequest),
+		providerRegChan:     rm.defaultProviders.providerRegChan,
+		cancel:              rm.defaultProviders.cancel,
+	}
+
+	newRm := &defaultProviderResmon{
```
Comment (Frassle) [inline pkg/resource/deploy/source_eval.go:2275]: probably call this a subcontextResmon or similar, it's unlikely to stay just for default providers.

File: pkg/resource/deploy/source_eval.go (hunk 5 above)
Comment (tgummerer) [reply pkg/resource/deploy/source_eval.go:458]: Nah, this is broken I think, and currently unused which is why I didn't noticed in testing.  I'll just remove this.

File: pkg/resource/deploy/source_eval.go (hunk 6 above)
Comment (tgummerer) [reply pkg/resource/deploy/source_eval.go:2275]: Yeah, I'll rename it :+1:

File: proto/pulumi/resource.proto (hunk 4 above)
Comment (tgummerer) [reply proto/pulumi/resource.proto]: Done :heavy_check_mark:

**Pull request closed without merging**